### PR TITLE
Improve error when parsing several licenses

### LINKED ISSUE #6103: Error trying to give multiple licenses
## Expected Behavior

https://dune.readthedocs.io/en/latest/dune-files.html?highlight=license#generate-opam-files-1 says:

> `(license <names>)` - specifies the license of the project, ideally as an identifier from the [SPDX License List](https://spdx.org/licenses/). Multiple licenses may be specified.

## Actual Behavior

I tried:

```
(license "ISC" "MIT")   # Error: Too many argument for license
```

```
(license ("ISC" "MIT"))  # Error: Atom or quoted string expected
```

```
(license "ISC")
(license "MIT")   # Error: Field "license" is present too many times
```

The following is accepted by dune, but is rejected by opam-repository (see https://github.com/ocaml/opam-repository/pull/18944/commits/d3ea851e9730368a1d74ebb18cf3a021922e34ac).

```
(license "(ISC AND MIT)")
```

It is possible to work around this using `PKG.opam.template` files.

## Specifications

```
$ dune --version
3.4.1

$ ocamlc --version
5.0.0~alpha1
```

### PR DESCRIPTION
**Pull request by emillon**
Closes #6103

### CODE REVIEW

**Comment by emillon** [comment]
I can land the refactoring commit separately if you prefer.

**Review by Alizter — APPROVED** [review]

File: src/dune_sexp/decoder.mli [hunk 1]
```diff
@@ -120,8 +120,9 @@ val kind : (kind, _) parser
     end of sequence is reached. *)
 val repeat : 'a t -> 'a list t
 
-(** Like [repeat] but the list of elements must be non-empty. *)
-val repeat1 : 'a t -> 'a list t
+(** Like [repeat] but the list of elements must be non-empty. [check_multi] is
+    performed when the parsed list has length > 1. *)
+val repeat1 : ?check_multi:unit t -> 'a t -> 'a list t
```
Comment (rgrinberg) [inline src/dune_sexp/decoder.mli:125]: why do we need to pass check_multi as an argument. Can't we always rewrite `repeat1 ~check_multi t` as:

```
let* x = repeat1 y in
let+ () = check_multi x in
x
```

Comment (emillon) [reply src/dune_sexp/decoder.mli:125]: the reason I pushed it into `repeat1` is that this avoids the need for `let*` in callers.
I see 2 benefits in this:
- these call sites stay simpler
- this leaves the door open to making `Decoder` purely applicative (`repeat1` would be a constructor)

Comment (rgrinberg) [reply src/dune_sexp/decoder.mli:125]: The call sites are a little shorter, but not simpler. One has to understand this new argument, and keep in mind when it's requireed. The trivial construct it replaces is much simpler than all that.

I'm not sure how much of a goal it is it to make the parser applicative, but it wouldn't be right to do it by adding new arguments to capture every behavior.

The right way to go would be to add a combinator that does what you want. Perhaps something like:

```
val ( <* ) : 'a t -> unit t -> 'a t
```

Comment (emillon) [reply src/dune_sexp/decoder.mli]: Allright, let's scrap that idea.

**Pull request merged**